grafana/grafana
# Frontend o11y: Report browser crashes to Faro

Fixes #94678

The implementation of the detection is in https://github.com/ifrost/crashme but can be moved to core or Faro if proves to be useful. The idea is to enable it in dev and ops and monitor it.

### How to test it?

* Check video [here](https://raintank-corp.slack.com/archives/C07JP5RU5KR/p1730721811683449)
* Open ephemeral instance: https://ephemeral1511182195772ifrost.grafana-dev.net/
* Paste the code snippet below to enable a toolbar to simulate tab crash
* Make sure the toggle is enabled (you can use the toolbar from the snippet to enable it)
* Simulate a crash using one of the buttons
* Check the crash report in the dashboard (link in the [Design Doc](https://docs.google.com/document/d/15y5adylEdV06aSEdWcg1i1FtN7NUW-ut3Sk4RTGVU7o/edit#heading=h.5sybau7waq2q))

<details><summary>Code snippet to crash the browser</summary>

Run below in dev tools or create a code snippet:

```javascript
let crashmeSimulation = {
  memoryCrash: function () {
    let dump = [];
    let dumps = {};

    setInterval(() => {
      for (let i = 0; i < 1000000; i++) {
        if (dump.length > 1000000) {
          dumps[Object.keys(dumps).length] = dump;
          dump = [];
        }
        dump.push(Math.random());
      }
    }, 1);
  },
  loopCrash: function () {
    let i = 0;
    while (true) {
      i++;
    }
  },
  recursiveCrash: function () {
    this.recursiveCrash();
    this.recursiveCrash();
    this.recursiveCrash();
  },
  domCrash: function () {
    const div = document.createElement('div');
    document.body.appendChild(div);
    while (true) {
      const child = div.cloneNode();
      document.body.appendChild(child);
    }
  },
};

let ui = document.createElement("div");
ui.style.position = "fixed";
ui.style.bottom = "20px"
ui.style.right = "20px";
ui.style.backgroundColor = "#333";
ui.style.padding = "5px";
ui.style.zIndex = 1000;

function makeButton(label, handler, options) {
  let button = document.createElement("button");
  button.addEventListener("click", handler);
  button.style.border = "1px solid #fff";
  button.style.color = "#ff";
  button.style.margin = "5px";
  button.style.padding = "5px";
  button.style.backgroundColor = options?.backgroundColor;
  button.innerText = label;
  return button;
}


function enableToggle() {
  localStorage.setItem("grafana.featureToggles", "crashDetection=true");
  location.reload();
}

function disableToggle() {
  localStorage.setItem("grafana.featureToggles", "crashDetection=false");
  location.reload();
}

const enabled = localStorage.getItem("grafana.featureToggles") === "crashDetection=true"

if (enabled) {
  ui.appendChild(makeButton("toggle enabled (click to disable)", () => disableToggle(), { backgroundColor: "green" }))
} else {
  ui.appendChild(makeButton("toggle disabled (click to enable)", () => enableToggle(), { backgroundColor: "red" }))
}

ui.appendChild(makeButton("OOM crash", () => crashmeSimulation.memoryCrash()))
ui.appendChild(makeButton("OOM crash 2", () => crashmeSimulation.domCrash()))
ui.appendChild(makeButton("Loop crash", () => crashmeSimulation.loopCrash()))
ui.appendChild(makeButton("Loop crash 2", () => crashmeSimulation.recursiveCrash()))


document.body.appendChild(ui);
```
</summary>
